src/numerics/include/metaphysicl/parallel_numberarray.h

```diff
@@ -0,0 +1,177 @@
+//-----------------------------------------------------------------------bl-
+//--------------------------------------------------------------------------
+//
+// MetaPhysicL - A metaprogramming library for physics calculations
+//
+// Copyright (C) 2013 The PECOS Development Team
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the Version 2.1 GNU Lesser General
+// Public License as published by the Free Software Foundation.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc. 51 Franklin Street, Fifth Floor,
+// Boston, MA  02110-1301  USA
+//
+//-----------------------------------------------------------------------el-
+
+#ifndef METAPHYSICL_PARALLEL_NUMBERARRAY_H
+#define METAPHYSICL_PARALLEL_NUMBERARRAY_H
+
+#include "metaphysicl/metaphysicl_config.h"
+
+#ifdef METAPHYSICL_HAVE_TIMPI
+
+#include "metaphysicl/numberarray.h"
+#include "timpi/standard_type.h"
+
+namespace TIMPI
+{
+template <std::size_t N, typename T>
+class StandardType<MetaPhysicL::NumberArray<N, T>> : public DataType
+{
+public:
+  explicit StandardType(
+      const MetaPhysicL::NumberArray<N, T> * example = nullptr)
+  {
+#ifdef TIMPI_HAVE_MPI
+    static data_type static_type = MPI_DATATYPE_NULL;
+    if (static_type == MPI_DATATYPE_NULL)
+      {
+        // We need an example for MPI_Address to use
+        static const MetaPhysicL::NumberArray<N, T> p {0};
+        if (!example)
+          example = &p;
+
+        static_assert(N > 0, "Zero-length NumberArray not supported by TIMPI");
+        StandardType<T> T_type(&((*example)[0]));
+
+        int blocklength = N;
+        MPI_Aint displs, start;
+        MPI_Datatype tmptype, type = T_type;
+
+        timpi_call_mpi
+          (MPI_Get_address (example, &start));
+        timpi_call_mpi
+          (MPI_Get_address (&((*example)[0]), &displs));
+
+        // subtract off offset to first value from the beginning of the structure
+        displs -= start;
+
+        // create a prototype structure
+        timpi_call_mpi
+          (MPI_Type_create_struct (1, &blocklength, &displs, &type,
+                                   &tmptype));
+        timpi_call_mpi
+          (MPI_Type_commit (&tmptype));
+
+        // resize the structure type to account for padding, if any
+        timpi_call_mpi
+          (MPI_Type_create_resized (tmptype, 0, sizeof(MetaPhysicL::NumberArray<N,T>),
+                                    &static_type));
+
+        timpi_call_mpi
+          (MPI_Type_free (&tmptype));
+
+        SemiPermanent::add
+          (std::make_unique<ManageType>(static_type));
```

Comment:
This is the code that ensures we don't leak, right?

Reply:
Right.